cmd/tendermint/commands/rollback.go

```diff
@@ -0,0 +1,109 @@
+package commands
+
+import (
+	"fmt"
+
+	"github.com/spf13/cobra"
+
+	"github.com/tendermint/tendermint/internal/state"
+	"github.com/tendermint/tendermint/version"
+)
+
+var RollbackStateCmd = &cobra.Command{
+	Use:   "rollback",
+	Short: "rollback tendermint state by one height",
+	Long: `
+Rollbacking state should be performed in the event of an incorrect applicate state transition
+whereby Tendermint has persisted an incorrect app hash and is thus unable to make 
+progress. This takes a state at height n and overwrites it with the state at height n - 1. 
+It is expected that the application also rolls back to height n - 1. No blocks are removed 
+so upon restarting Tendermint, the transactions in block n will be reexecuted against the 
+application.
+`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		// use the parsed config to load the block and state store
+		bs, ss, err := loadStateAndBlockStore(config)
+		if err != nil {
+			return err
+		}
+
+		// rollback the last state
+		return rollbackState(bs, ss)
+	},
+}
+
+// rollbackState takes a state at height n and overwrites it with the state
+// at height n - 1. Note state here refers to tendermint state not application state.
+func rollbackState(bs state.BlockStore, ss state.Store) error {
```

Comment:
Could you make this function public?
And maybe place it in a different package?

In any case, an sdk command should be able to call this and then do it's own rollback

Reply:
+1 to making this be somewhere else and maybe testing writing some more unittests.